flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/asyncprocessing/AbstractAsyncStateStreamOperator.java

```diff
@@ -149,6 +149,25 @@ public final void preserveRecordOrderAndProcess(ThrowingRunnable<Exception> proc
         asyncExecutionController.syncPointRequestWithCallback(processing);
     }
 
+    @Override
+    @SuppressWarnings("unchecked")
+    public <K> void asyncProcessWithKey(K key, ThrowingRunnable<Exception> processing) {
+        RecordContext<K> previousContext = currentProcessingContext;
+
+        // build a context and switch to the new context
+        currentProcessingContext = asyncExecutionController.buildContext(null, key);
+        currentProcessingContext.retain();
+        asyncExecutionController.setCurrentContext(currentProcessingContext);
+        // Same logic as RECORD_ORDER, since FIRST_STATE_ORDER is problematic when the call's key
```

Comment:
Should we call `super.setCurrentKey()` for `stateHandler` here? 
FLINK-36096 needs to get the current key from the state backend.

Reply:
I think instead of setting current key here, we should set current key during any context switch.

This should be reconsidered, and before that, I'd prefer keeping here as it is.

Reply:
Thanks for the clarification, OK